File: algorithms/2_medium/M/01654_minimum_jumps_to_reach_home/bfs.cpp
constexpr int MAX_ALLOWED = 2'000;

class Solution {
public:
    static int minimumJumps(const vector<int>& forbidden, const int a, const int b, const int x) {
        assert(x >= 0 && a >= 1 && b >= 1 && a <= MAX_ALLOWED && b <= MAX_ALLOWED && x <= MAX_ALLOWED);

        if (x == 0) { return 0; }

        int seen_len = 0;
        if (a >= b) {
            seen_len = x + b + 1;
        } else {
            const int max = *max_element(forbidden.cbegin(), forbidden.cend());
            seen_len = std::max(max, x) + a + b + 1;
        };

        auto seen = vector<pair<bool, bool>>(seen_len, make_pair(false, false));

        for (const int f : forbidden) {
            assert(f >= 1 && f != x);
            if (f < seen_len) {
                seen[f] = make_pair(true, true);
            }
        }

        auto q = deque<pair<int16_t, bool>>();
        q.push_back(make_pair(0, false));
        seen[0] = make_pair(true, true);
        int turns = 0;

        while (!q.empty()) {
            const size_t curr_len = q.size();
            for (size_t steps = 0; steps < curr_len; steps += 1)  {
                const auto [pos_i16, can_back_jump] = q.front();
                q.pop_front();
                const int pos = pos_i16;

                if (pos == x) {
                    return turns;
                }

                if (pos + a < seen_len && !seen[pos + a].first) {
                    seen[pos + a] = make_pair(true, true);
                    q.push_back(make_pair(static_cast<int16_t>(pos + a), true));
                }

                if (can_back_jump && pos - b > 0 && !seen[pos - b].second) {
                    seen[pos - b].second = true;
                    q.push_back(make_pair(static_cast<int16_t>(pos - b), false));
                }
            }

            turns += 1;
        }

        return -1;
    }
};

File: algorithms/2_medium/M/01654_minimum_jumps_to_reach_home/bfs.rs
use std::collections::VecDeque;

const MAX_ALLOWED: i32 = 2_000;

impl Solution {
    pub fn minimum_jumps(forbidden: Vec<i32>, a: i32, b: i32, x: i32) -> i32 {
        assert!(x >= 0 && a >= 1 && b >= 1 && a <= MAX_ALLOWED && b <= MAX_ALLOWED && x <= MAX_ALLOWED);

        if x == 0 { return 0; }

        let seen_len = if a >= b {
            x + b + 1
        } else {
            let max = *forbidden.iter().max().unwrap();
            max.max(x) + a + b + 1
        };

        let mut seen = vec![(false, false); seen_len as usize];

        for f in forbidden {
            assert!(f >= 1 && f != x);
            if f < seen_len {
                seen[f as usize] = (true, true);
            }
        }

        let mut q: VecDeque<(i16, bool)> = VecDeque::new();
        q.push_back((0, false));
        seen[0] = (true, true);
        let mut turns = 0;

        while !q.is_empty() {
            let curr_len = q.len();
            for _ in 0..curr_len  {
                let (pos, can_back_jump) = q.pop_front().unwrap();
                let pos = pos as i32;

                if pos == x {
                    return turns;
                }

                if pos + a < seen_len && !seen[(pos + a) as usize].0 {
                    seen[(pos + a) as usize] = (true, true);
                    q.push_back(((pos + a) as i16, true));
                }

                if can_back_jump && pos - b > 0 && !seen[(pos - b) as usize].1 {
                    seen[(pos - b) as usize].1 = true;
                    q.push_back(((pos - b) as i16, false));
                } 
            }

            turns += 1;
        }

        -1
    }
}
